core/tests/test_storage.cache.py

```diff
@@ -50,6 +50,51 @@ def test_get_set(self):
         with self.assertRaises(RuntimeError):
             cache.get(KEY)
 
+    def test_decorator_mismatch(self):
+        with self.assertRaises(AssertionError):
+
+            @cache.stored(KEY)
+            async def async_fun():
+                pass
+
+    def test_decorators(self):
+        run_count = 0
+        cache.start_session()
+
+        @cache.stored(KEY)
+        def func():
+            nonlocal run_count
+            run_count += 1
+            return "foo"
+
+        # cache is empty
+        self.assertIsNone(cache.get(KEY))
+        self.assertEqual(run_count, 0)
+        self.assertEqual(func(), "foo")
+        # function was run
+        self.assertEqual(run_count, 1)
+        self.assertEqual(cache.get(KEY), "foo")
+        # function does not run again but returns cached value
+        self.assertEqual(func(), "foo")
+        self.assertEqual(run_count, 1)
+
+        @cache.stored_async(KEY)
+        async def async_func():
+            nonlocal run_count
+            run_count += 1
+            return "bar"
+
+        # cache is still full
+        self.assertEqual(await_result(async_func()), "foo")
```

Comment:
That would be safer, yes, but probably not worth the complication when interacting with the unimport mechanism.

Namely: imagine a Cardano flow, which uses the caching decorator on `apps.cardano.seed.get_keychain`. The decorator gets called when the flow runs, and locks the `APP_CARDANO_ROOT` entry. When the next Cardano action happens, the Cardano app is imported again, the decorator runs again and raises an exception because the cache entry is already in use.

We would need a way to identify functions across imports and reimports, and I'm not sure if there is a way to do that.

Reply:
Ah, I see. Hm I agree it sounds like too much trouble.